astropy/coordinates/transformations.py

```diff
@@ -674,6 +674,108 @@ def deco(func):
+
+        Parameters
+        ----------
+        fromsys : class
+            The coordinate frame class to start from.
+        viasys : class
+            The coordinate frame class to transform through.
+        tosys : class
+            The coordinate frame class to transform into.
+        priority : number
+            The priority of this transform when finding the shortest
+            coordinate transform path - large numbers are lower priorities.
+
+        Notes
+        -----
+        Even though the created transform is a single step in the graph, it
+        will still internally call the constituent transforms.  Thus, there is
+        no performance benefit for using this created transform.
+
+        For Astropy's built-in frames, the most common choice for ``viasys`` is
+        `~astropy.coordinates.ICRS`.
+
+        An error will be raised if a direct transform between ``fromsys`` and
+        ``tosys`` already exists, regardless of whether that existing transform
+        internally uses ``viasys``.
+        """
+        full_path = self.get_transform(fromsys, tosys)
+        if full_path is None:
+            raise ValueError(f"{fromsys.__name__}->{tosys.__name__} is not possible")
+        if len(full_path.transforms) == 1:
+            raise ValueError(f"A direct transform for {fromsys.__name__}->{tosys.__name__} already exists")
+
+        transform_string = f"{fromsys.__name__}->{tosys.__name__} via {viasys.__name__}"
+
+        first_path = self.get_transform(fromsys, viasys)
+        second_path = self.get_transform(viasys, tosys)
+        if first_path is None or second_path is None:
+            raise ValueError(f"{transform_string} is not possible")
+
+        trans_class_set = set([type(t) for t in first_path.transforms] +
+                              [type(t) for t in second_path.transforms])
+
+        # Shared frame attributes are used for instantiating the intermediate frame
+        # For any attributes that are shared by the "from"/"to" frames, the "to" frame takes precedence
+        second_overlap = set(second_path.overlapping_frame_attr_names)
+        first_overlap = set(first_path.overlapping_frame_attr_names) - second_overlap
+
+        def direct_get_via_attrs(from_coo, to_frame):
+            via_attrs = dict(zip(first_overlap, [getattr(from_coo, n) for n in first_overlap]))
+            via_attrs.update(zip(second_overlap, [getattr(to_frame, n) for n in second_overlap]))
+            return via_attrs
+
+        # If all of the transforms are affine, we can create an affine transform
+        if all([issubclass(tc, BaseAffineTransform) for tc in trans_class_set]):
+
+            # If all of the transforms do not return offsets, we can use DynamicMatrixTransform
+            if all([issubclass(tc, (StaticMatrixTransform, DynamicMatrixTransform))
+                    for tc in trans_class_set]):
+                def direct_DMT(from_coo, to_frame):
```

Comment:
This and the below functions don't need to be defined inside of this method. They might be best as private funcs defined outside. At least for readability, when trying to understand this method, the internals of ``direct_DMT``, etc are not crucial.

Reply:
Another suggestion would be to have these as `staticmethod` on `DynamicMatrixTransform` and `AffineTransform`. But definitely do not let the perfect be the enemy of good enough!

Reply:
I think these review comments missed a subtlety.  These transformation functions are not generic functions, but rather are transformation-path-specific functions that access transformation-path-specific variables that are not passed in through their call signatures.  It's not possible to simply define them outside as is.  In principal, one could define generic functions and then dynamically create the necessary transformation-path-specific functions using `partial()`, but that would become significantly less readable.

Please take a look at current version post-refactoring